### DISCUSSION TITLE
Error while initializing the PMW3901 sensor after soldering directly to Crazyflie 2.1

### DISCUSSION THREAD

**Discussion by andrew-t97** [bitcraze/discussions#1568] [Q&A]
I’m having some trouble with the Flow deck V2 on my Crazyflie 2.1. I’ve directly soldered the Flow deck to my Crazyflie. Unfortunately, I’m now getting issues with the deck’s initialisation. It seems to be an issue with the PMW3901. I’ve checked the voltages of each of its related components, and I believe they’re okay. I also can’t see any physical damage to the deck. Does anyone know what might be causing this behaviour? Could I have missed something in the soldering/voltage measurement?

The firmware I’m using is built from the master branch of the firmware repository at commit `7f1f78251fccdb355d672d579997238258e367d2` (24th of July 2024). I’ve added the logs I get in the cfclient and the voltages I measured below. I’ve also attached some images of the Crazyflie2.1 and the deck (I’m sorry you have to witness my soldering 😭).

**Logs**

```
SYS: ----------------------------
SYS: Crazyflie 2.1 is up and running!
SYS: Build 169:7f1f78251fcc (2023.11 +169) CLEAN
SYS: I am 0x313232313232471200230024 and I have 1024KB of flash!
CFGBLK: v1, verification [OK]
DECK_CORE: 1 deck(s) found
DECK_CORE: Calling INIT on driver bcFlow2 for deck 0
ZR2: Z-down sensor [OK]
PMW: Motion chip id: 0x:0x
IMU: BMI088: Using I2C interface.
IMU: BMI088 Gyro connection [OK].
IMU: BMI088 Accel connection [OK]
IMU: BMP388 I2C connection [OK]
ESTIMATOR: Using Kalman (2) estimator
CONTROLLER: Using PID (1) controller
MTR-DRV: Using brushed motor driver
SYS: About to run tests in system.c.
SYS: NRF51 version: 2024.2 (CF21)
EEPROM: I2C connection [OK].
STORAGE: Storage check [OK].
IMU: BMI088 gyro self-test [OK]
Error while initializing the PMW3901 sensor
DECK_CORE: Deck 0 test [FAIL].
SYS: deck [FAIL]
```

**Voltages**

C7 - 1.8
C8 - 3
C9 - 1.7
C12 - 3
C13 - 1.7  

R7 -  1.8 / 1.8 
R8 - 3 / 2.85

Thanks so much for your time and any help you can provide!

![PXL_20240929_135057928](https://github.com/user-attachments/assets/e0281734-9662-449a-a4f0-b3fed8624ccd)
![PXL_20240929_135210654](https://github.com/user-attachments/assets/58b36d86-8e58-472f-a9df-9b553ae64a64)

**Comment by tobbeanton** [comment]
The PMW3901 is using the SPI bus so I would start by double checking those connections (PA5, PA6, PA7, PB4). The supply voltages looks fine.

**Reply by andrew-t97** [reply]
Hi @tobbeanton, I've checked the voltages on the Crazyflie 2.1 board. Here are the results below. I'm assuming that PA5, PA6, and PA7 should not be zero. 😕

**Voltages**

PA5 - 0
PA6 - 0
PA7 - 0
PB4 - 3

**Comment by tobbeanton** [comment]
I think the voltages are fine. The SPI bus is running in mode 0 and then the signals will be low at idle. The measurement does not tell if the signal is running intact all the way to the PMW3901. Do you have a scope to measure it with during a transaction such as boot-up?

**Reply by andrew-t97** [reply]
Okay. Unfortunately I don't have a scope. ☹️ I could try changing the wires to confirm that it's not a dodgy wire. Is there anything else I can try to diagnose the issue without a scope?

**Reply by tobbeanton** [reply]
You could use the multi-meter in resistance mode and try to check that all signals are intact. You can also measure the voltages during boot on PA5, PA6, and PA7 to see that you get more then 0.

**Reply by andrew-t97** [reply]
I'll give that a go! In terms of checking if the signals are intact, what terminals do I need to measure in resistance mode? Sorry, I'm quite new to debugging electronics. 😄

**Reply by andrew-t97** [reply]
I've rewired the deck but still haven't gotten a reading on PA5, 6, or 7. I've tried measuring the voltages during boot (while the rotors are doing self-checks), and I still get zero on PA5, 6, and 7. 🤔

**Reply by tobbeanton** [reply]
> In terms of checking if the signals are intact, what terminals do I need to measure in resistance mode? 

The more of the signal path you can check the better, but start with the wires.

> I still get zero on PA5, 6, and 7. 🤔

It is most likely that the communication is so short the multimeter does not have time detect it. A scope or logic analyser is then needed.

**Reply by andrew-t97** [reply]
I've checked the resistance of the wires and all check out as zero. 🎉 I've ordered a logic analyser and will be able to check PA5, 6, and 7 later today. Will the logic analyser essentially tell me whether the issue is with the Crazyflie sending the SPI signals or the PMW3901 receiving them?

**Reply by andrew-t97** [reply]
Hi @tobbeanton, so I've done some further debugging. I wasn't having much luck with the logic analyser (no readings on SPI), so I ordered a new flow deck to see if that was the problem. 📦

When I attach the flow deck to the pins on the top of the crazyflie, the flow deck boots up correctly (no PMW3901 error). ✅ However, when I connect the deck to the wires I've soldered onto the bottom of the deck, I get the PMW3901 error. I'm assuming there's some issue with the SPI connection on the base. I haven't removed the plastic pin blocks on the top of the crazyflie for soldering. Instead, I've just soldered the wires directly to the bottom. Do you reckon this might be the issue? Should I remove the pin block from the top? 🤔

**Reply by tobbeanton** [reply]
Wires could cause problem with high speed communication but if I recall correctly the PMW3901 is only running on 2MHZ.

> I wasn't having much luck with the logic analyser (no readings on SPI)

That is strange and might point to the problem. Did you get a reading when you used the working deck?

> Instead, I've just soldered the wires directly to the bottom. Do you reckon this might be the issue?

I don't think this should be an issue if the soldering is ok.

**Reply by andrew-t97** [reply]
So I made another attempt and got the crazyflie to boot with the new flow deck! However, my crazyflie has now decided not to boot anymore. 😭 I've created a discussion [here](https://github.com/orgs/bitcraze/discussions/1587) for it. I'm wondering if the old flow deck's PMW3901 was somehow damaged, although the lack of SPI readings was weird. If I can get my crazyflie working again, I'll follow this up 🤞
